tests/core/test_featurizer.py

```diff
@@ -109,3 +112,67 @@ def test_label_tokenizer_featurizer_handles_probabilistic_intents():
         {"intent_a": 0.5, "prev_b": 0.2, "intent_d": 1.0, "prev_action_listen": 1.0}
     )
     assert (encoded == np.array([0.5, 1.0, 1.5, 0.0, 0.2])).all()
+
+def test_bow_tokenizer_featurizer_handles_on_non_existing_features():
+    f = BOWSingleStateFeaturizer()
+    d = Domain
+    d.input_states = ['ab_cd', 'ef_gh_kl']
+    f.prepare_from_domain(d)
+    encoded = f.encode({"ab_cd": 1.0, "prev_cd_ab": 0.0, "kl": 1.0, "prev_action_listen": 1.0}
+        )
+    assert (encoded == np.array([2, 2, 0, 0, 1])).all()
+
+def test_bow_tokenizer_featurizer_uses_correct_dtype_int():
+    f = BOWSingleStateFeaturizer()
+    d = Domain
+    d.input_states = ['ab_cd', 'ef_gh_kl']
+    f.prepare_from_domain(d)
+    encoded = f.encode({"ab_cd": 1.0, "prev_cd_ab": 0.0, "kl": 1.0, "prev_action_listen": 1.0})
+    assert encoded.dtype == np.int32
+
+def test_bow_tokenizer_featurizer_outputs_numpy_as_required():
+    f = BOWSingleStateFeaturizer()
+    d = Domain
+    d.input_states = ['ab_cd', 'ef_gh_kl']
+    f.prepare_from_domain(d)
+    encoded = f.encode({"ab_cd": 1.0, "prev_cd_ab": 0.0, "kl": 1.0, "prev_action_listen": 1.0})
+    assert type(encoded) == np.ndarray
+
+def test_bow_tokenizer_featurizer_outputs_sparse_matrix_as_required():
+    f = BOWSingleStateFeaturizer()
+    d = Domain
+    d.input_states = ['ab_cd', 'ef_gh_kl']
+    f.prepare_from_domain(d)
+    encoded = f.encode({"ab_cd": 1.0, "prev_cd_ab": 0.0, "kl": 1.0, "prev_action_listen": 1.0}, type_output="sparse")
+    assert type(encoded) == scipy.sparse.csr.csr_matrix
+
+def test_bow_tokenizer_featurizer_none_input_sparse():
+    f = BOWSingleStateFeaturizer()
+    d = Domain
+    d.input_states = ['ab_cd', 'ef_gh_kl']
+    f.prepare_from_domain(d)
+    encoded = f.encode(None, type_output="sparse")
+    assert type(encoded) == scipy.sparse.csr.csr_matrix
+
+def test_bow_tokenizer_featurizer_none_input_dense():
+    f = BOWSingleStateFeaturizer()
+    d = Domain
+    d.input_states = ['ab_cd', 'ef_gh_kl']
+    f.prepare_from_domain(d)
+    encoded = f.encode(None, type_output="dense")
+    assert type(encoded) == np.ndarray and encoded.dtype == np.int32
+
+def main():
+    test_bow_tokenizer_featurizer_handles_on_non_existing_features()
+    test_bow_tokenizer_featurizer_uses_correct_dtype_int()
+    test_bow_tokenizer_featurizer_outputs_numpy_as_required()
+    test_bow_tokenizer_featurizer_outputs_sparse_matrix_as_required()
+    test_bow_tokenizer_featurizer_none_input_sparse()
+    test_bow_tokenizer_featurizer_none_input_dense()
+
+
+main()
```

Comment:
pycharm let's you run tests by clicking on the green arrow next to the test or right-clicking on the module. Other IDEs should have similar features